reference-implementation/lib/writable-stream.js

```diff
@@ -182,215 +190,248 @@ function WritableStreamAddWriteRequest(stream) {
-    if (wasAborted === true) {
+    if (stream._pendingAbortRequest !== undefined) {
       stream._pendingAbortRequest._reject(stream._storedError);
       stream._pendingAbortRequest = undefined;
     }
 
-    WritableStreamRejectPromisesInReactionToError(stream);
+    WritableStreamRejectPromisesInReactionToError(stream, stream._storedError);
 
     return;
   }
 
-  const controller = stream._writableStreamController;
+  assert(state === 'writable');
 
-  if (wasAborted === false) {
+  if (stream._pendingAbortRequest === undefined) {
     return;
   }
 
   WritableStreamFinishAbort(stream, state);
 
   const abortRequest = stream._pendingAbortRequest;
   stream._pendingAbortRequest = undefined;
-  const promise = WritableStreamDefaultControllerAbort(controller, abortRequest._reason);
+  const promise = WritableStreamDefaultControllerAbort(stream._writableStreamController, abortRequest._reason);
   promise.then(
     abortRequest._resolve,
     abortRequest._reject
   );
 }
 
-function WritableStreamFinishPendingWriteWithError(stream, reason) {
-  assert(stream._pendingWriteRequest !== undefined);
-  stream._pendingWriteRequest._reject(reason);
-  stream._pendingWriteRequest = undefined;
+function WritableStreamFinishInFlightWriteWithError(stream, reason) {
+  assert(stream._inFlightWriteRequest !== undefined);
+  stream._inFlightWriteRequest._reject(reason);
+  stream._inFlightWriteRequest = undefined;
 
   const state = stream._state;
 
-  let wasAborted = false;
-  if (stream._pendingAbortRequest !== undefined) {
-    wasAborted = true;
-  }
+  if (state === 'errored') {
+    if (stream._pendingAbortRequest !== undefined) {
+      stream._pendingAbortRequest._reject(stream._storedError);
+      stream._pendingAbortRequest = undefined;
+    }
 
-  let readyPromiseIsPending = false;
-  if (state === 'writable' && wasAborted === false &&
-      WritableStreamDefaultControllerGetBackpressure(stream._writableStreamController) === true) {
-    readyPromiseIsPending = true;
-  }
+    WritableStreamRejectPromisesInReactionToError(stream, stream._storedError);
 
-  if (wasAborted === true) {
-    stream._pendingAbortRequest._reject(reason);
-    stream._pendingAbortRequest = undefined;
+    return;
   }
 
-  if (state === 'errored') {
-    WritableStreamRejectPromisesInReactionToError(stream);
+  assert(state === 'writable');
+
+  if (stream._pendingAbortRequest === undefined) {
+    stream._state = 'errored';
+    stream._storedError = reason;
+
+    const writer = stream._writer;
+    if (writer !== undefined) {
+      if (WritableStreamCloseInQueueOrInFlight(stream) === false && stream._backpressure === true) {
+        defaultWriterReadyPromiseReject(writer, reason);
+      } else {
+        defaultWriterReadyPromiseResetToRejected(writer, reason);
+      }
+      writer._readyPromise.catch(() => {});
+    }
+
+    WritableStreamRejectPromisesInReactionToError(stream, reason);
 
     return;
   }
 
   stream._state = 'errored';
```

Comment:
I don't understand why this two lines are not shared with line 249 above (ie. moved above the loop condition).